src/QMCDrivers/DMC/WalkerControl.cpp

```diff
@@ -236,13 +208,15 @@ void WalkerControl::branch(int iter, MCPopulation& pop, bool do_not_branch)
     throw std::runtime_error("Potential bug! Population num_global_walkers mismatched!");
 #endif
 
+  // This defensive code implies that previous code left population walkers in invalid state.
```

Comment:
This is not a defensive code.

Reply:
Ok so by my reading through the walker life cycle, branching, and population control walkers must always have Multiplicty == 1.0 after amplification and/or transfer. Multiplicity is set in a small number of significant locations (except for here). 

1.  On construction the default member initializer of `Walker` sets the initial value to 1.0
2.  During warmup = Multiplicity = 1.0
3.  To `static_cast<int>(walker->Weight + rng_())` resulting in a integer value of >= 0
4.  After swapping where on rank multiplicity is set to the remaining number of copies on a rank.
5. After swapping where a new walker has its multiplicity set to the number of copies received + 1 (number of copies received is counted from 0 due to lack of discrimination between counting numbers and indexing numbers).
6.  After amplificiation i.e. `copyHighMultiplicityWalkers` and `killDeadWalkers` the population will contain only walkers  of Multiplicity == 1.
7. Here where all living walkers have their multiplicity set to 1.

Several off these state transforms appear defensive in nature only, i.e. they can only cover up Walkers arriving to this function in an invalid state.

2. Can only erase a state transformation to walker that left Multiplicity in an incorrect state. Fresh walkers have multiplicity == 1 walkers from walker configuration have multiplicity == 1. An assert should be added and this side of the branch is noop.
7. If multiplicity != something is wrong, this simply covers it.

4, 5 are only necessary because of how we construct transfer jobs. Where basically walker have a multiplicity per destination for a short span of code. During this section Walker->Multiplicty is no long a simple value.

Multiplicity is a very important value for correct population control so its update and validity so reasoning about its state changes and validity should be as simple as possible.

Reply:
I'm clarifying the comments, I don't want to lose this insight but it can't really be cleaned up properly until we drop the legacy drivers.